go/vt/key/key.go

```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+	return a.Start
 }
 
-// KeyRangeEndEqual returns true if both key ranges have the same end
-func KeyRangeEndEqual(left, right *topodatapb.KeyRange) bool {
-	if left == nil {
-		return right == nil || len(right.End) == 0
+func keyRangeEndMin(a, b *topodatapb.KeyRange) []byte {
+	if Empty(a.End) {
+		return b.End
+	}
+	if Empty(b.End) {
+		return a.End
 	}
-	if right == nil {
-		return len(left.End) == 0
+	if Less(a.End, b.End) {
+		return a.End
 	}
-	return bytes.Equal(addPadding(left.End), addPadding(right.End))
+	return b.End
 }
 
-// For more info on the following functions, see:
-// See: http://stackoverflow.com/questions/4879315/what-is-a-tidy-algorithm-to-find-overlapping-intervals
-// two segments defined as (a,b) and (c,d) (with a<b and c<d):
-// intersects = (b > c) && (a < d)
-// overlap = min(b, d) - max(c, a)
-
-// KeyRangesIntersect returns true if some Keyspace values exist in both ranges.
-func KeyRangesIntersect(first, second *topodatapb.KeyRange) bool {
-	if first == nil || second == nil {
-		return true
+// KeyRangesOverlap returns a KeyRange representing the overlap between KeyRange a and b. If they do not overlap,
+// an error is returned.
+func KeyRangesOverlap(a, b *topodatapb.KeyRange) (*topodatapb.KeyRange, error) {
+	if !KeyRangesIntersect(a, b) {
+		return nil, fmt.Errorf("KeyRanges %v and %v don't overlap", a, b)
+	}
+	if KeyRangeIsComplete(a) {
+		return b, nil
+	}
+	if KeyRangeIsComplete(b) {
+		return a, nil
 	}
-	return (len(first.End) == 0 || bytes.Compare(second.Start, first.End) < 0) &&
-		(len(second.End) == 0 || bytes.Compare(first.Start, second.End) < 0)
-}
 
-// KeyRangesOverlap returns the overlap between two KeyRanges.
-// They need to overlap, otherwise an error is returned.
-func KeyRangesOverlap(first, second *topodatapb.KeyRange) (*topodatapb.KeyRange, error) {
-	if !KeyRangesIntersect(first, second) {
-		return nil, fmt.Errorf("KeyRanges %v and %v don't overlap", first, second)
-	}
-	if first == nil {
-		return second, nil
-	}
-	if second == nil {
-		return first, nil
-	}
-	// compute max(c,a) and min(b,d)
-	// start with (a,b)
-	result := proto.Clone(first).(*topodatapb.KeyRange)
-	// if c > a, then use c
-	if bytes.Compare(second.Start, first.Start) > 0 {
-		result.Start = second.Start
-	}
-	// if b is maxed out, or
-	// (d is not maxed out and d < b)
-	//                           ^ valid test as neither b nor d are max
-	// then use d
-	if len(first.End) == 0 || (len(second.End) != 0 && bytes.Compare(second.End, first.End) < 0) {
-		result.End = second.End
-	}
-	return result, nil
+	return &topodatapb.KeyRange{Start: keyRangeStartMax(a, b), End: keyRangeEndMin(a, b)}, nil
 }
 
-// KeyRangeIncludes returns true if the first provided KeyRange, big,
-// contains the second KeyRange, small. If they intersect, but small
-// spills out, this returns false.
+// KeyRangeIncludes returns true if KeyRange big fully contains KeyRange small.
 func KeyRangeIncludes(big, small *topodatapb.KeyRange) bool {
```

Comment:
Currently that function name is already used for `KeyRangeContains(keyRange *topodatapb.KeyRange, id []byte)` and Go doesn't support method overloading. It could be named `KeyRangeContainsKeyRange (a, b *topodatapb.KeyRange)` though! What do you think?